src/extensions/scratch3_video_sensing/lib.js

```diff
@@ -0,0 +1,275 @@
+/**
+ * lib.js
+ *
+ * Tony Hwang and John Maloney, January 2011
+ * Michael "Z" Goddard, March 2018
+ *
+ * Video motion sensing primitives.
+ */
+
+const TO_DEGREE = 180 / Math.PI;
+const WIDTH = 480;
+const HEIGHT = 360;
+// chosen empirically to give a range of roughly 0-100
+const AMOUNT_SCALE = 100;
+// note 2e-4 * activePixelNum is an experimentally tuned threshold for my
+// logitech Pro 9000 webcam - TTH
+const LOCAL_AMOUNT_SCALE = AMOUNT_SCALE * 2e-4;
+const THRESHOLD = 10;
+const WINSIZE = 8;
+const LOCAL_MAX_AMOUNT = 100;
+const LOCAL_THRESHOLD = THRESHOLD / 3;
+
+const STATE_KEY = 'Scratch.videoSensing';
+
+class VideoMotion {
+    constructor () {
+        this.frameNumber = 0;
+        this.motionAmount = 0;
+        this.motionDirection = 0;
+        this.analysisDone = false;
+
+        this.curr = null;
+        this.prev = null;
+
+        this._arrays = new ArrayBuffer(WIDTH * HEIGHT * 2 * 1);
+        this._curr = new Uint8ClampedArray(this._arrays, WIDTH * HEIGHT * 0 * 1, WIDTH * HEIGHT);
+        this._prev = new Uint8ClampedArray(this._arrays, WIDTH * HEIGHT * 1 * 1, WIDTH * HEIGHT);
+    }
+
+    reset () {
+        this.prev = this.curr = null;
+        this.motionAmount = this.motionDirection = 0;
+        this.analysisDone = true;
+
+        const targets = this.runtime.targets;
+        for (let i = 0; i < targets.length; i++) {
+            targets[i].getCustomState(STATE_KEY).motionAmount = 0;
+            targets[i].getCustomState(STATE_KEY).motionDirection = 0;
+        }
+    }
+
+    addFrame (source) {
+        this.frameNumber++;
+
+        this.prev = this.curr;
+        this.curr = new Uint32Array(source.buffer.slice());
+
+        const _tmp = this._prev;
+        this._prev = this._curr;
+        this._curr = _tmp;
+        for (let i = 0; i < this.curr.length; i++) {
+            this._curr[i] = this.curr[i] & 0xff;
+        }
+
+        this.analysisDone = false;
+    }
+
+    analyzeFrame () {
+        if (!this.curr || !this.prev) {
+            this.motionAmount = this.motionDirection = -1;
+            // don't have two frames to analyze yet
+            return;
+        }
+
+        const winStep = (WINSIZE * 2) + 1;
+        const wmax = WIDTH - WINSIZE - 1;
+        const hmax = HEIGHT - WINSIZE - 1;
+
+        let i;
+        let j;
+        let address;
+
+        let A2;
+        let A1B2;
+        let B1;
+        let C1;
+        let C2;
+        let u;
+        let v;
+        let uu;
+        let vv;
+        let n;
```

Comment:
I'd suggest moving these inside the relevant scopes, where possible. For example, it looks like the `A`/`B`/`C` family could move into the `j` loop, to where they get initialized each loop. Unless JS is silly enough to cause a performance hit in a case like that...?

Reply:
I think that'll be better for readability.